### ISSUE TITLE
SQLAlchemy.__init__ stopped working for create_model in 1.5.4

### ISSUE THREAD

**Issue by killthekitten** [flask-admin/flask-admin#1927]
The custom `__init__` does not get called when the model is created through the `create_view`. Noticed after upgrading 1.5.3->1.5.4.

#1811 seems to introduce the bug, and the [follow-on patch](https://github.com/flask-admin/flask-admin/commit/731d14c16f57debd31f7f7f394d1585e20c28d5d 
) doesn't help.

Tested with SQLAlchemy 1.3.8 and 1.3.11 by throwing a breakpoint in the `__init__`.

**Comment by killthekitten** [comment]
Easy way to reproduce: https://gist.github.com/killthekitten/e9c2549a22bd8a372c2164781a9d524d

**Comment by etcsayeghr** [comment]
I can confirm that I'm also experiencing this bug. Will pin down to 1.5.3 in hopes that resolves our issue.

**Comment by mrjoes** [comment]
I'm not sure that this is a bug, more here: https://docs.sqlalchemy.org/en/13/orm/constructors.html

Long story short, SQLAlchemy does not call `__init__` when it creates new instances of a model because `__init__` might have a custom signature. Flask-Admin 1.5.4 follows the same convention now - whenever the model is instantiated, it is following the SQLAlchemy convention of not calling `__init__` on the model. Try adding the `@orm.reconstructor` decorator and see if it will help.

**Comment by killthekitten** [comment]
I can see why the change was introduced in the first place, but I'd still consider this a bug.

My use case for `__init__` is to set defaults for a new record, most of the time. Whereas `@orm.reconstructor` serves a different purpose: 
> The SQLAlchemy ORM does not call `__init__` when recreating objects from database rows. ... If you need to do some setup on database-loaded instances before they’re ready to use, ... it is also available via a class-specific decorator called `orm.reconstructor()`. 

Explicit `__init__` [is ok in the declarative](https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/basic_use.html#class-constructor) SQLAlchemy, and there are [examples](https://docs.sqlalchemy.org/en/13/orm/tutorial.html#building-a-many-to-many-relationship) scattered all over the docs. So the update introduces quite a bit of headache for everyone who uses `__init__`.

One way to resolve this on my side would be to change all models to use the `default=` argument in the column declarations, but there's nothing fundamentally wrong about the current code, and I'm sure I will end up writing an explicit init again in the future. Because, heh, this is python.

Or I could indeed maintain a patch, or more likely not update to a newer version. 

Is there anything we can do to `Flask-Admin` support all cases without any changes to the user code?

**Comment by mrjoes** [comment]
I'm definitely open to suggestions, but there was a problem before where people would do something like this:

```
class MyModel(BaseModel):
    def __init__(self, arg1, arg2):
        ...
```

This is a valid SQLAlchemy model, it'll work just fine because SQLAlchemy does not call a default constructor when the model is instantiated when it's read/constructed from a database.

But this also prevented Flask-Admin from creating the model because it was creating its instance without passing any arguments to a constructor as there's no way for Flask-Admin to know what to pass in this scenario.

Basically, yes, you can have custom `__init__` constructors in your models, but SQLAlchemy would never use them. And Flask-Admin would never use them too.

I'm not sure that this is a best scenario, but it kind of unified model constructor behavior between SQLAlchemy and Flask-Admin.

On a side note, there's no need to downgrade to 1.5.3 as it is possible to create a base class with overridden `create_model` method that will have 1.5.3 behavior by creating a class instance in a generic way and then inheriting all project model views from it.

**Comment by killthekitten** [comment]
Sounds good, I will add a doc entry based on your comments then.

**Comment by killthekitten** [comment]
@mrjoes I looked into it once again. It seems a lot of boilerplate to override the `create_model`. Perhaps, we could add a hook (or a view level config) to build a model instance?

```python
def model_factory(self):
    return self.model()

def create_model(self, form):
        try:
            model = self.model_factory()
            form.populate_obj(model)
            model.put()
        except Exception as ex:
            if not self.handle_view_exception(ex):
                # flash(gettext('Failed to create record. %(error)s',
                #    error=ex), 'error')
                logging.exception('Failed to create record.')
            return False
        else:
            self.after_model_change(form, model, True)

        return model
```

**Comment by killthekitten** [comment]
@mrjoes I'm following up on the discussion with #1981. It introduces a custom hook that you can override directly, without copying the entire `create_model` into the base class.

**Comment by mrjoes** [comment]
Yep, thanks for the MR - accepted.
